### PR TITLE
Speak typed characters in windows consoles in the correct language

### LINKED ISSUE #10113: UIA console: when typing in the console, characters are wrongly pronounced regardless of the keyboard layout
### Steps to reproduce:
1. open a console window with the uia console option turned on, for example a wsl console
. switch to croatian or to the US keyboard language. I am not aware of other layouts having the problem which i will describe here thoroughly
russian keyboard is also included there
2. try to type  or ], ; or ', \ or /
### Actual behavior:
note: here are given the corresponding keys on the us keyboard. these keys are wrongly interpreted by NVDA also on oother non-us keyboard layout
for example / slash on US keyboard is interpreted as apostrophe,
; as č, the croatian diacritic mark, and so on.
### Expected behavior:
the signs should be interpreded bz NVDA as these are written on the kezboard respectivelz
#### NVDA installed/portable/running from source:
NVDA installed
#### NVDA version:
Latest alpha
#### Windows version:
18362 latest indows 10 1903
#### Name and version of other software in use when reproducing the issue:
n-a
#### Other information about your system:
to reproduce this bug, you should have turned on have character echo turned on
#### Does the issue still occur after restarting your PC?
yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
n/a

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #10113 

### Summary of the issue:
With the work to support Windows consoles via UIA, NVDA moved to speaking typed characters in consoles by detecting keyboard input and then using ToUnicodeEx to convert the keyboard input into a character for speaking. We had already been taking this approach for Microsoft Edge and other places in Windows 10 where we could not inject in-process.
ToUnicodeEx requires a keyboard layout to use the correct language. Therefore, we would call GetKeyboardLayout, giving it the thread ID of the focused window.
However, it seems that Windows console windows report the wrong thread ID, and therefore, GetKeyboardLayout fails, returning 0 instead of a valid keyboard layout.
Passing this 0 to ToUnicodeEx, it would generate characters in the user's system language (I think). At least, on my system it was always english US no matter what input language I was currently using.
 
### Description of how this pull request fixes the issue:
We now try multiple things to get a valid keyboard layout:
* Firstly we try getting it from the focus as usual.
* If that fails, we instead use the keyboard layout detected from the last input language change for the foreground window. This keyboard layout is now cached on the foreground object when detected by nvdaControlerInternal_inputLangChange in NVDAHelper.py.
* Finally, if there is no cached keyboard layout, we fallback to calling GetKeyboardLayout for NVDA's main thread. 

### Testing performed:
* Installed the Arabic keyboard layout.
* Opened a Windows console and ensured NVDA was using UIA.
* With speak typed characters on, first typed some characters, confirming they were being announced in english. backspacing them also confirming they were english characters written to the console.
* Switched keyboard layouts to Arabic.
* Again typed some characters, confirming that NVDA was speaking them as Arabic. Backspacing them, confirmed that actual arabic characters were written to the console.
  
### Known issues with pull request:
None.

### Change log entry:
None

### CODE REVIEW

File: source/NVDAHelper.py [hunk 1]
```diff
@@ -336,6 +336,11 @@ def nvdaControllerInternal_inputLangChangeNotify(threadID,hkl,layoutString):
 	#Simple case where there is no change
 	if languageID==lastLanguageID and layoutString==lastLayoutString:
 		return 0
+	fg = api.getForegroundObject()
+	# Cache the new keyboard layout on the foreground object,
+	# Which may be used by keyboardHandler
+	# when speaking typed characters in situations where it cannot be fetched directly from the focus.
+	fg._lastDetectedKeyboardLayoutChange = hkl
```
Comment (leonardder) [inline source/NVDAHelper.py]: I'd suggest storing this in globalVars instead.

File: source/keyboardHandler.py [hunk 2]
```diff
@@ -219,7 +219,19 @@ def internal_keyDownEvent(vkCode,scanCode,extended,injected):
...
+			if not hkl:
+				log.debug("Failed to fetch keyboard layout from focus, trying layout from last detected change")
+				# Some threads, such as for Windows consoles
+				# Do not allow getKeyboardLayout to work.
+				# Therefore, use the cached keyboard layout from the last inputLangChange detected by NVDA
+				# on the foreground object.
+				hkl = getattr(api.getForegroundObject(), '_lastDetectedKeyboardLayoutChange', 0)
+				if not hkl:
+					log.debug("No layout cached, falling back to layout of NVDA main thread")
+					# As a last resort, use the keyboard layout of NVDA's main thread.
```
Comment (leonardder) [inline source/keyboardHandler.py]: How likely would it be that the keyboard layout for the NVDA main thread differs from the keyboard layout of the currently focused app?

**Comment by michaelDCurran** [comment]
In Windows 7, input language changes were per GUI thread.

In Windows 10, the default is now that input language changes apply 
across the entire desktop. However, this can be configured to apply 
per-app again, in Settings -> System -> time and language -> keyboard 
language settings.


If that option didn't exist, we could just use the thread of the shell 
window, or perhaps NVDA's main thread -- though it does seem that NvDA's 
UI needs to get focus before it picks up a desktop-wide input language 
change.

Since it can be per-app, we need to track keyboard layouts per thread, 
or at very least, per app (process ID).


Now thinking about it, I think it will be safe enough to cache it on the 
appModule.


However, the real problem here is that the inputLangChange notification 
for console windows comes from a conhost thread, where as Windows has 
hacked ConsoleWindowClass windows to reflect the thread and processID of 
the attached process requesting input (I.e. Ubuntu bash). We could work 
out the conhost processID from the thread ID and then get the appModule 
for that and cache the layout there. But, the other day I created / 
merged a pr that deliberately forced the processID property for 
winConsole UIA NVDAObject's to again reflect the process ID of the 
hosted app, not conhost. These two ideas are in conflict with each other.

I'll have to think about this further...

**Comment by leonardder** [comment]
Thanks for the explanation. Because it's problematic, that's why I suggested globalVars. We only have to cache just one layout anyway. My personal belief is that caching things on objects should be avoided.

**Comment by michaelDCurran** [comment]
But if we cache on globalVars, then we can only track one layout. And if the user has configured for per-app input language changes, this will cause us to sometimes use the wrong layout again.

I think the better approach here is going to be finding a way to get the real input thread of conhost and passing that to GetKeyboardLayout like we do for Edge and such.

**Comment by codeofdusk** [comment]
This may also fix the dead key bug mentioned in #9915.

**Comment by michaelDCurran** [comment]
Replaced by pr #10554

**Pull request closed without merging**